## Сингулярное разложение матриц

1. Найдите посредством NumPy SVD для матрицы:

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

**Решение**

In [2]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

In [7]:
a = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])
print(f'Матрица A: \n{a}')

Матрица A: 
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [4]:
u, s, w = np.linalg.svd(a)

In [5]:
d = np.zeros_like(a, dtype=float)
d[np.diag_indices(min(a.shape))] = s

In [8]:
print(f'Матрица D: \n{d}')

Матрица D: 
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


In [9]:
print(f'Матрица U: \n{u}')

Матрица U: 
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]


In [10]:
print(f'Матрица V: \n{w.T}')

Матрица V: 
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


Сделаем проверку:

In [13]:
print(np.dot(np.dot(u, d), w))

[[ 1.  2. -0.]
 [-0.  0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [ 0.  1.  0.]]


2. Для матрицы из предыдущего задания найдите:

    а) евклидову норму;
    
    б) норму Фробениуса.

**Решение**

а) Евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел. Максимальное значение полученного отношения будет равно максимальному сингулярному числу $\mu_{max}$:

$$\left \| A \right \|_{E}=8.82$$.

Проверим с помощью numpy:

In [14]:
print(f'Евклидова норма: {np.linalg.norm(a, ord=2)}')

Евклидова норма: 8.824868854820442


б) Напишем функцию для нахождения нормы Фробениуса:

In [17]:
def frobenius_func(matrix):
    sum_square = 0
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            sum_square += matrix[i][j] ** 2
    return np.sqrt(sum_square)

print(f'Норма фробениуса для матрицы A:\n{frobenius_func(a)}')

Норма фробениуса для матрицы A:
11.045361017187261


Проверим с помощью numpy:

In [19]:
print(f"Норма Фробениуса: \n{np.linalg.norm(a, ord='fro')}")

Норма Фробениуса: 
11.045361017187261
